### PR TITLE
fix: KES signer period

### PR DESCRIPTION
**Pull request by jpraynaud**
## Content

This PR includes a **fix to use the current KES period as input to the KES signer when signing instead of the the number of evolutions of the KES secret key since creation**. This avoids circular dependency where the Operational certificate (returned by the KES signer) is needed to compute the correct number of evolutions of the KES secret key. 
It also includes some clarification on the KES evolutions used to verify the KES signature.

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] All check jobs of the CI have succeeded
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

## Issue(s)
Relates to #2833

### CODE REVIEW

File: mithril-common/src/test/crypto_helper/cardano/kes/setup.rs [hunk 1]
```diff
@@ -31,16 +31,12 @@ pub fn create_kes_cryptographic_material(
...
     let mut dummy_buffer = [0u8; Sum6Kes::SIZE + 4];
     let mut dummy_seed = [party_idx as u8; 32];
-    let (mut kes_secret_key, kes_verification_key) =
+    let (kes_secret_key, kes_verification_key) =
         Sum6Kes::keygen(&mut dummy_buffer, &mut dummy_seed);
-    for _ in 0..kes_period {
-        kes_secret_key
-            .update()
-            .expect("KES secret key update should not fail");
-    }
```
Comment (Alenar) [inline mithril-common/src/test/crypto_helper/cardano/kes/setup.rs:40]: For my understanding, what does this change entails ?

File: mithril-signer/src/runtime/runner.rs [hunk 2]
```diff
@@ -184,7 +184,7 @@ impl Runner for SignerRunner {
         let stake = stake_distribution
             .get(&self.services.single_signer.get_party_id())
             .ok_or_else(RunnerError::NoStakeForSelf)?;
-        let (operational_certificate, protocol_operational_certificate) = match &self
+        let (_operational_certificate, protocol_operational_certificate) = match &self
```
Comment (Alenar) [inline mithril-signer/src/runtime/runner.rs]: This could be further simplified :)
```suggestion
        let protocol_operational_certificate = match &self
```

File: mithril-common/src/entities/signer.rs [hunk 3]
```diff
@@ -158,8 +158,10 @@ pub struct SignerWithStake {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub operational_certificate: Option<ProtocolOpCert>,
 
-    /// The kes period used to compute the verification key signature
-    // TODO: This kes period should not be used as is and should probably be within an allowed range of kes period for the epoch
+    /// The number of evolutions of the KES key since the start KES period of the operational certificate at the time of signature.
+    ///
+    /// Note: the naming 'kes_period' lacks clarity and should be renamed to 'kes_evolutions'
+    // TODO: This 'kes_period' should be rename to 'kes_evolutions' to avoid confusion
```
Comment (Alenar) [inline mithril-common/src/entities/signer.rs]: I'm confused, the place in the code were this value is assigned is in the signer runtime, the value now assigned is:
```rust
let kes_period = self.services.chain_observer.get_current_kes_period().await?;
``` 
Before it was:
```rust
let kes_period = match operational_certificate {
    Some(operational_certificate) => Some(
        self.services
            .chain_observer
            .get_current_kes_period()
            .await?
            .unwrap_or_default()
            - operational_certificate.get_start_kes_period() as KesPeriod,
        ),
        None => None,
    };
```
I may be missing something but this is no KES evolutions, it was the evolution before the change of this PR in the signer runtime but now it seems to be the KES period at the time of the signer creation.

Also this change is far from neutral and also impact messages, the impact is low because we don't really use this value (it's ignored by the aggregator verifier), it's mostly informative.

**Review by Alenar — COMMENTED** [review]
Maybe we should introduce two new types: `KesPeriod` and `KesEvolution`, the two would be wrappers like we do for the `BlockNumber`.
There could be a twist: the implementation of the `sub` operators on `KesPeriod` could result in `KesEvolution`.

File: mithril-common/src/test/crypto_helper/cardano/kes/setup.rs (hunk 1 above)
Comment (jpraynaud) [reply mithril-common/src/test/crypto_helper/cardano/kes/setup.rs:40]: These updates do not have any impact on the serialized secret key, which explains why they have been removed.

File: mithril-common/src/entities/signer.rs (hunk 3 above)
Comment (jpraynaud) [reply mithril-common/src/entities/signer.rs]: You're right, good catch :+1:  

I have fixed that by using the `let current_kes_period = self.services.chain_observer.get_current_kes_period().await?;` to build the protocol initializer and keep unchanged the `kes_evolutions` to create the `Signer` entity.

File: mithril-common/src/entities/signer.rs [hunk 4]
```diff
@@ -158,10 +158,10 @@ pub struct SignerWithStake {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub operational_certificate: Option<ProtocolOpCert>,
 
-    /// The kes period used to compute the verification key signature
-    // TODO: This kes period should not be used as is and should probably be within an allowed range of kes period for the epoch
+    /// The number of evolutions of the KES key since the start KES period of the operational certificate at the time of signature.
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub kes_period: Option<KesPeriod>,
+    #[serde(rename = "kes_period")]
```
Comment (Alenar) [inline mithril-common/src/entities/signer.rs]: You should be able to combine the two attribute :)
```suggestion
    #[serde(rename = "kes_period", skip_serializing_if = "Option::is_none")]
```

File: mithril-common/src/messages/message_parts/signer.rs [hunk 5]
```diff
@@ -36,11 +36,10 @@ pub struct SignerWithStakeMessagePart {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub operational_certificate: Option<HexEncodedOpCert>,
 
-    /// The KES period used to compute the verification key signature
-    // TODO: This KES period should not be used as is and should probably be
-    //       within an allowed range of KES periods for the epoch.
+    /// The number of evolutions of the KES key since the start KES period of the operational certificate at the time of signature.
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub kes_period: Option<KesPeriod>,
+    #[serde(rename = "kes_period")]
```
Comment (Alenar) [inline mithril-common/src/messages/message_parts/signer.rs]:
```suggestion
    #[serde(rename = "kes_period", skip_serializing_if = "Option::is_none")]
```

File: mithril-common/src/messages/message_parts/signer.rs [hunk 6]
```diff
@@ -166,11 +165,10 @@ pub struct SignerMessagePart {
...
```
Comment (Alenar) [inline mithril-common/src/messages/message_parts/signer.rs]:
```suggestion
    #[serde(rename = "kes_period", skip_serializing_if = "Option::is_none")]
```

File: mithril-common/src/messages/register_signer.rs [hunk 7]
```diff
@@ -34,11 +34,10 @@ pub struct RegisterSignerMessage {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub operational_certificate: Option<HexEncodedOpCert>,
 
-    /// The KES period used to compute the verification key signature
-    // TODO: This KES period should not be used as is and should probably be
-    //       within an allowed range of KES periods for the epoch.
+    /// The number of evolutions of the KES key since the start KES period of the operational certificate at the time of signature.
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub kes_period: Option<KesPeriod>,
+    #[serde(rename = "kes_period")]
```
Comment (Alenar) [inline mithril-common/src/messages/register_signer.rs]:
```suggestion
    #[serde(rename = "kes_period", skip_serializing_if = "Option::is_none")]
```

File: mithril-common/src/entities/signer.rs (hunk 4 above)
Comment (jpraynaud) [reply mithril-common/src/entities/signer.rs]: Fixed

File: mithril-common/src/messages/message_parts/signer.rs (hunk 5 above)
Comment (jpraynaud) [reply mithril-common/src/messages/message_parts/signer.rs]: Fixed

File: mithril-common/src/messages/message_parts/signer.rs (hunk 6 above)
Comment (jpraynaud) [reply mithril-common/src/messages/message_parts/signer.rs]: Fixed

File: mithril-common/src/messages/register_signer.rs (hunk 7 above)
Comment (jpraynaud) [reply mithril-common/src/messages/register_signer.rs]: Fixed

File: mithril-common/src/crypto_helper/cardano/kes/verifier_standard.rs [hunk 8]
```diff
@@ -19,19 +19,19 @@ impl KesVerifier for KesVerifierStandard {
...
     ) -> StdResult<()> {
         operational_certificate
             .validate()
             .map_err(|_| KesVerifyError::OpCertInvalid)?;
 
-        // Check if the KES period in the operational certificate matches the provided KES period +/- 1
-        let kes_period_try_min = std::cmp::max(0, kes_period.saturating_sub(1));
-        let kes_period_try_max = std::cmp::min(64, kes_period.saturating_add(1));
-        for kes_period_try in kes_period_try_min..kes_period_try_max {
+        // Check if the signature verifies for the provided KES evolutions value +/- 1
```
Comment (curiecrypt) [inline mithril-common/src/crypto_helper/cardano/kes/verifier_standard.rs]: Should we also test valid/invalid cases for +-1 evolutions in the verification?
I cannot see a test that covers these edge cases

File: mithril-common/src/entities/signer.rs [hunk 9]
```diff
@@ -71,7 +71,7 @@ impl Signer {
             verification_key,
             verification_key_signature,
             operational_certificate,
-            kes_period,
+            kes_evolutions: kes_period,
```
Comment (curiecrypt) [inline mithril-common/src/entities/signer.rs]: I would rather rename the function argument as `kes_evolutions`

File: mithril-common/src/entities/signer.rs [hunk 10]
```diff
@@ -200,7 +199,7 @@ impl SignerWithStake {
...
```
Comment (curiecrypt) [inline mithril-common/src/entities/signer.rs]: same as above

File: mithril-common/src/messages/message_parts/signer.rs [hunk 11]
```diff
@@ -256,7 +252,7 @@ impl Debug for SignerMessagePart {
                     "operational_certificate",
                     &format_args!("{:?}", self.operational_certificate),
                 )
-                .field("kes_period", &format_args!("{:?}", self.kes_period))
+                .field("kes_period", &format_args!("{:?}", self.kes_evolutions))
```
Comment (curiecrypt) [inline mithril-common/src/messages/message_parts/signer.rs]: Is it possible to rename the field as well? 
```suggestion
                .field("kes_evolutions", &format_args!("{:?}", self.kes_evolutions))
```

File: mithril-aggregator/src/database/record/signer_registration.rs [hunk 12]
```diff
@@ -25,7 +25,10 @@ pub struct SignerRegistrationRecord {
     /// Operational certificate of the stake pool operator associated to the signer
     pub operational_certificate: Option<HexEncodedOpCert>,
 
-    /// The kes period used to compute the verification key signature
+    /// The number of evolutions of the KES key since the start KES period of the operational certificate at the time of signature.
+    ///
+    /// Note: the naming 'kes_period' lacks clarity and should be renamed to 'kes_evolutions'
+    // TODO: This 'kes_period' should be renamed to 'kes_evolutions' to avoid confusion
```
Comment (curiecrypt) [inline mithril-aggregator/src/database/record/signer_registration.rs:31]: +1

File: mithril-common/src/crypto_helper/cardano/kes/signer_with_key.rs [hunk 13]
```diff
@@ -61,85 +65,114 @@ mod tests {
...
         let KesCryptographicMaterialForTest {
             party_id: _,
             operational_certificate_file,
             kes_secret_key_file,
         } = create_kes_cryptographic_material(
             1 as KesPartyIndexForTest,
-            0 as KesPeriod,
-            "create_valid_signature_for_message",
+            start_kes_period,
+            current_function!(),
```
Comment (curiecrypt) [inline mithril-common/src/crypto_helper/cardano/kes/signer_with_key.rs:85]: Just out of curiosity... Why do we need this `current_function()` now and not in the previous version?

Comment (jpraynaud) [reply mithril-common/src/crypto_helper/cardano/kes/signer_with_key.rs:85]: This macro returns the name of the current function and we use it to create a sub-folder for the cryptographic material.
This avoids having multiple use the same sub-folder by mistake which usually creates flakiness in the CI when tests are run in parallel.

File: mithril-common/src/entities/signer.rs (hunk 9 above)
Comment (jpraynaud) [reply mithril-common/src/entities/signer.rs]: Fixed

File: mithril-common/src/entities/signer.rs (hunk 10 above)
Comment (jpraynaud) [reply mithril-common/src/entities/signer.rs]: Fixed

File: mithril-common/src/messages/message_parts/signer.rs (hunk 11 above)
Comment (jpraynaud) [reply mithril-common/src/messages/message_parts/signer.rs]: Fixed

File: mithril-common/src/crypto_helper/cardano/kes/verifier_standard.rs (hunk 8 above)
Comment (jpraynaud) [reply mithril-common/src/crypto_helper/cardano/kes/verifier_standard.rs]: Yes, I added a test for that and also fixed a problem with the range which was too narrow.
